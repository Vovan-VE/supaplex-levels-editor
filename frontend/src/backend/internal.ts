import type { StoreDriver, StoreDriverSingle } from "@cubux/storage-driver";
import { CodeOf } from "@cubux/types";

export type ConfigStorage = StoreDriverSingle<string, any>;

export type FilesStorageKey = CodeOf<"FilesStorageItem">;
export interface FilesStorageItem {
  key: FilesStorageKey;
  fileBuffer: ArrayBuffer;
}
export type FilesStorage = StoreDriver<string, FilesStorageItem>;

export interface OpenFileItem {
  file: File;
  key?: FilesStorageKey;
}
export type OpenFiles = readonly OpenFileItem[];
export type OpenFileDoneCallback = (items: OpenFiles) => void;
export interface OpenFileOptions {
  multiple?: boolean;
  done: OpenFileDoneCallback;
}
