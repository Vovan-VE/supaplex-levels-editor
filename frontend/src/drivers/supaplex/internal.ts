import { IBounds, Rect } from "utils/rect";
import {
  IsPlayableResult,
  ITilesRegion,
  IWithDemo,
  IWithSignature,
} from "../types";

export interface ISupaplexBox extends IBounds {
  readonly length: number;
  coordsToOffset(x: number, y: number): number;
  validateCoords?(x: number, y: number): void;
}

export interface ILevelBody extends ITilesRegion {
  readonly length: number;
  readonly raw: Uint8Array;
  setTile(x: number, y: number, value: number): ILevelBody;
  batch(update: (b: ILevelBody) => ILevelBody): ILevelBody;
  isPlayable(): IsPlayableResult;
  copyRegion(r: Rect): readonly [x: number, y: number, region: ILevelBody];
  findPlayer(): [x: number, y: number] | null;
}

export const enum GravityStatic {
  OFF = 0,
  ON = 1,
}
export const enum FreezeZonksStatic {
  OFF = 0,
  ON = 2,
}
export const enum FreezeEnemiesStatic {
  OFF = 0,
  ON = 1,
}
export const enum SpecPortAlterMod {
  NOTHING = -1,
  TOGGLE = -2,
}
export type SpecPortAlterRaw = number & {};
export type SpecPortAlter = SpecPortAlterMod | SpecPortAlterRaw;
export type SpecPortGravity = GravityStatic | SpecPortAlter;
export type SpecPortFreezeZonks = FreezeZonksStatic | SpecPortAlter;
export type SpecPortFreezeEnemies = FreezeEnemiesStatic | SpecPortAlter;

export interface ISupaplexSpecPortStringOptions {
  withZeros?: boolean;
}
export interface ISupaplexSpecPortsIO {
  isStdCompatible(width: number): boolean;
  toString(o?: ISupaplexSpecPortStringOptions): string;
  toRaw(width: number): Uint8Array;
}
export interface ISupaplexSpecPortRecordReadonly {
  readonly x: number;
  readonly y: number;
  //TODO: ? readonly ex: number;
  //TODO: ? readonly ey: number;
  readonly gravity: SpecPortGravity;
  readonly freezeZonks: SpecPortFreezeZonks;
  readonly freezeEnemies: SpecPortFreezeEnemies;
  readonly unusedByte: number;
}
export interface ISupaplexSpecPortRecord
  extends ISupaplexSpecPortRecordReadonly,
    ISupaplexSpecPortsIO {
  setX(x: number): ISupaplexSpecPortRecord;
  setY(y: number): ISupaplexSpecPortRecord;
  setGravity(v: SpecPortGravity): ISupaplexSpecPortRecord;
  setFreezeZonks(v: SpecPortFreezeZonks): ISupaplexSpecPortRecord;
  setFreezeEnemies(v: SpecPortFreezeEnemies): ISupaplexSpecPortRecord;
  setUnusedByte(u: number): ISupaplexSpecPortRecord;
}
export interface ISupaplexSpecPortDatabase extends ISupaplexSpecPortsIO {
  readonly count: number;
  readonly countStdCompatible: number;
  getAll(): Iterable<ISupaplexSpecPortRecord>;
  copySpecPortsInRegion(r: Rect): ISupaplexSpecPortDatabase;
  clear(): ISupaplexSpecPortDatabase;
  find(x: number, y: number): ISupaplexSpecPortRecord | null;
  add(x: number, y: number): ISupaplexSpecPortDatabase;
  set(p: ISupaplexSpecPortRecord): ISupaplexSpecPortDatabase;
  update(
    x: number,
    y: number,
    update: (prev: ISupaplexSpecPortRecord) => ISupaplexSpecPortRecord,
  ): ISupaplexSpecPortDatabase;
  delete(x: number, y: number): ISupaplexSpecPortDatabase;
}

export interface ILevelFooter extends IWithDemo, IWithSignature {
  readonly length: number;
  getRaw(): Uint8Array;
  readonly title: string;
  setTitle(title: string): this;
  readonly initialGravity: boolean;
  setInitialGravity(on: boolean): this;
  readonly initialFreezeZonks: boolean;
  setInitialFreezeZonks(on: boolean): this;
  readonly infotronsNeed: number;
  setInfotronsNeed(value: number): this;
  readonly usePlasma: boolean;
  readonly usePlasmaLimit: number | undefined;
  readonly usePlasmaTime: number | undefined;
  readonly useZonker: boolean;
  readonly useSerialPorts: boolean;
  readonly useInfotronsNeeded: number | undefined;
  readonly initialFreezeEnemies: boolean;
  setUsePlasma(on: boolean): this;
  setUsePlasmaLimit(n: number | undefined): this;
  setUsePlasmaTime(n: number | undefined): this;
  setUseZonker(on: boolean): this;
  setUseSerialPorts(on: boolean): this;
  setUseInfotronsNeeded(n: number | undefined): this;
  setInitialFreezeEnemies(on: boolean): this;
}

export const enum LocalOpt {
  UsePlasma = "usePlasma",
  UsePlasmaLimit = "usePlasmaLimit",
  UsePlasmaTime = "usePlasmaTime",
  UseZonker = "useZonker",
  UseSerialPorts = "useSerialPorts",
  UseInfotronsNeeded = "useInfotronsNeeded",
  PortsDatabase = "portsDB",
  InitialFreezeEnemies = "initialFreezeEnemies",
}
