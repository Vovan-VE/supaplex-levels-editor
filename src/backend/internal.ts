import type { StoreDriver, StoreDriverSingle } from "@cubux/storage-driver";
import { CodeOf } from "@cubux/types";

export type ConfigStorage = StoreDriverSingle<string, any>;

export type FilesStorageKey = CodeOf<"FilesStorageItem">;
export interface FilesStorageItem {
  key: FilesStorageKey;
  fileBuffer: ArrayBuffer;
}
export type FilesStorage = StoreDriver<string, FilesStorageItem>;

export interface OpenFileDoneItem {
  file: File;
  key?: FilesStorageKey;
}
export type OpenFileDoneCallback = (items: readonly OpenFileDoneItem[]) => void;
export interface OpenFileOptions {
  multiple?: boolean;
  done: OpenFileDoneCallback;
}
