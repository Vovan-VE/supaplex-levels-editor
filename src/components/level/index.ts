export * from "./body";
export * from "./toolbars";
