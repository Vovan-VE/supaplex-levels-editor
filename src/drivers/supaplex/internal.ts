import { IBounds, RectA } from "utils/rect";
import { IsPlayableResult, ITilesRegion, ITilesStreamItem } from "../types";

export interface ISupaplexBox extends IBounds {
  readonly length: number;
  coordsToOffset(x: number, y: number): number;
  validateCoords?(x: number, y: number): void;
}

export interface ILevelBody extends IBounds {
  readonly length: number;
  readonly raw: Uint8Array;
  getTile(x: number, y: number): number;
  setTile(x: number, y: number, value: number): ILevelBody;
  batch(update: (b: ILevelBody) => ILevelBody): ILevelBody;
  isPlayable(): IsPlayableResult;
  tilesRenderStream(
    x: number,
    y: number,
    w: number,
    h: number,
  ): Iterable<ITilesStreamItem>;
  copyRegion(
    x: number,
    y: number,
    w: number,
    h: number,
  ): readonly [x: number, y: number, region: ITilesRegion];
}

export interface ISupaplexSpecPortProps {
  setsGravity: boolean;
  setsFreezeZonks: boolean;
  setsFreezeEnemies: boolean;
}

export interface ISupaplexSpecPort extends ISupaplexSpecPortProps {
  x: number;
  y: number;
}

export interface ILevelFooter {
  readonly length: number;
  getRaw(): Uint8Array;
  readonly title: string;
  setTitle(title: string): this;
  readonly initialGravity: boolean;
  setInitialGravity(on: boolean): this;
  readonly initialFreezeZonks: boolean;
  setInitialFreezeZonks(on: boolean): this;
  readonly infotronsNeed: number;
  setInfotronsNeed(value: number): this;
  readonly specPortsCount: number;
  getSpecPorts(): Iterable<ISupaplexSpecPort>;
  copySpecPortsInRegion(r: RectA): readonly ISupaplexSpecPort[];
  clearSpecPorts(): this;
  findSpecPort(x: number, y: number): ISupaplexSpecPortProps | undefined;
  setSpecPort(x: number, y: number, props?: ISupaplexSpecPortProps): this;
  deleteSpecPort(x: number, y: number): this;
}
