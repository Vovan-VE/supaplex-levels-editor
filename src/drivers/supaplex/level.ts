import {
  ISupaplexLevel,
  ISupaplexSpecPort,
  ISupaplexSpecPortProps,
} from "./types";
import { isSpecPort } from "./tiles";

export const LEVEL_BYTES_LENGTH = 1536;
const LEVEL_WIDTH = 60;
const LEVEL_HEIGHT = 24;

const GRAVITY_OFFSET = 1444;
const TITLE_OFFSET = GRAVITY_OFFSET + 2;
const TITLE_LENGTH = 23;
const FREEZE_ZONKS_OFFSET = TITLE_OFFSET + TITLE_LENGTH;
const INFOTRONS_NEED_OFFSET = FREEZE_ZONKS_OFFSET + 1;
const SPEC_PORT_COUNT_OFFSET = INFOTRONS_NEED_OFFSET + 1;
const SPEC_PORT_DB_OFFSET = SPEC_PORT_COUNT_OFFSET + 1;
const SPEC_PORT_ITEM_LENGTH = 6;
const SPEC_PORT_MAX_COUNT = 10;

const coordsToOffset = (x: number, y: number) => y * LEVEL_WIDTH + x;

const specPortRawZeros = () => new Uint8Array(SPEC_PORT_ITEM_LENGTH);
const specPortRawOffset = (index: number) =>
  SPEC_PORT_DB_OFFSET + index * SPEC_PORT_ITEM_LENGTH;

export const specPortOffsetToCoords = (
  b0: number,
  b1: number,
): [x: number, y: number] => {
  const offset = ((b0 << 8) | b1) >> 1;
  return [offset % LEVEL_WIDTH, Math.floor(offset / LEVEL_WIDTH)];
};

export const specPortCoordsToOffset = (
  x: number,
  y: number,
): [b0: number, b1: number] => {
  const offset = coordsToOffset(x, y) << 1;
  return [(offset >> 8) & 0xff, offset & 0xff];
};

const getSpecPortProps = (raw: Uint8Array): ISupaplexSpecPortProps => ({
  setsGravity: raw[2] === 1,
  setsFreezeZonks: raw[3] === 2,
  setsFreezeEnemies: raw[4] === 1,
});

const validateCoords =
  process.env.NODE_ENV === "production"
    ? undefined
    : (x: number, y: number) => {
        if (x < 0 || x >= LEVEL_WIDTH || y < 0 || y >= LEVEL_HEIGHT) {
          throw new RangeError(`Cell coords (${x}, ${y}) are out of range`);
        }
      };

const validateByte =
  process.env.NODE_ENV === "production"
    ? undefined
    : (byte: number) => {
        if (byte < 0 || byte > 255) {
          throw new RangeError(`Invalid byte ${byte}`);
        }
      };

export class SupaplexLevel implements ISupaplexLevel {
  static fromArrayBuffer(buffer: ArrayBufferLike, byteOffset?: number) {
    return new SupaplexLevel(
      new Uint8Array(buffer, byteOffset, LEVEL_BYTES_LENGTH),
    );
  }

  readonly #data: Uint8Array;

  constructor(data?: Uint8Array) {
    if (data) {
      if (
        process.env.NODE_ENV !== "production" &&
        data.length !== LEVEL_BYTES_LENGTH
      ) {
        throw new Error(
          `Invalid buffer length ${data.length}, expected ${LEVEL_BYTES_LENGTH}`,
        );
      }
      this.#data = new Uint8Array(data);
    } else {
      this.#data = new Uint8Array(LEVEL_BYTES_LENGTH);
      this.title = "";
    }
  }

  copy() {
    return new SupaplexLevel(this.#data);
  }

  get raw() {
    return new Uint8Array(this.#data);
  }

  get width() {
    return LEVEL_WIDTH;
  }

  get height() {
    return LEVEL_HEIGHT;
  }

  get isResizable() {
    return false;
  }

  getCell(x: number, y: number) {
    validateCoords?.(x, y);
    return this.#data[coordsToOffset(x, y)];
  }

  setCell(x: number, y: number, value: number) {
    validateCoords?.(x, y);
    validateByte?.(value);
    const offset = coordsToOffset(x, y);
    if (isSpecPort(value)) {
      if (!isSpecPort(this.#data[offset])) {
        this.setSpecPort(x, y);
      }
    } else {
      if (isSpecPort(this.#data[offset])) {
        this.deleteSpecPort(x, y);
      }
    }
    this.#data[offset] = value;
  }

  get initialGravity() {
    return this.#data[GRAVITY_OFFSET] === 1;
  }
  set initialGravity(on) {
    this.#data[GRAVITY_OFFSET] = on ? 1 : 0;
  }

  get title() {
    return String.fromCharCode(
      ...this.#data.slice(TITLE_OFFSET, TITLE_OFFSET + TITLE_LENGTH),
    );
  }
  set title(title) {
    if (process.env.NODE_ENV !== "production" && title.length > TITLE_LENGTH) {
      throw new RangeError("Title length exceeds limit");
    }
    if (/[^\x20-\x7F]/.test(title)) {
      throw new Error("Unsupported characters found");
    }

    this.#data.set(
      title
        .padEnd(TITLE_LENGTH)
        .split("")
        .map((ch) => {
          const byte = ch.charCodeAt(0);
          validateByte?.(byte);
          return byte;
        }),
      TITLE_OFFSET,
    );
  }

  get maxTitleLength() {
    return TITLE_LENGTH;
  }

  get initialFreezeZonks() {
    return this.#data[FREEZE_ZONKS_OFFSET] === 2;
  }
  set initialFreezeZonks(on) {
    this.#data[FREEZE_ZONKS_OFFSET] = on ? 2 : 0;
  }

  get infotronsNeed() {
    return this.#data[INFOTRONS_NEED_OFFSET] || "all";
  }
  set infotronsNeed(inf) {
    if (inf === "all") {
      this.#data[INFOTRONS_NEED_OFFSET] = 0;
    } else {
      validateByte?.(inf);
      if (process.env.NODE_ENV !== "production" && inf === 0) {
        throw new RangeError("Value cannot be 0 since it has special meaning");
      }
      this.#data[INFOTRONS_NEED_OFFSET] = inf;
    }
  }

  get specPortsCount() {
    return Math.min(SPEC_PORT_MAX_COUNT, this.#data[SPEC_PORT_COUNT_OFFSET]);
  }

  #setSpecPortsCount(count: number) {
    this.#data[SPEC_PORT_COUNT_OFFSET] = count;
  }

  *getSpecPorts(): Generator<ISupaplexSpecPort, void, void> {
    for (let i = 0, L = this.specPortsCount; i < L; i++) {
      const raw = this.#getSpecPortRaw(i);
      const [x, y] = specPortOffsetToCoords(raw[0], raw[1]);
      yield {
        x,
        y,
        ...getSpecPortProps(raw),
      };
    }
  }

  clearSpecPorts() {
    let count = this.specPortsCount;
    this.#setSpecPortsCount(0);
    while (count-- > 0) {
      this.#data.set(specPortRawZeros(), specPortRawOffset(count));
    }
  }

  #findSpecPortOffset(x: number, y: number): number | undefined {
    const [b0, b1] = specPortCoordsToOffset(x, y);
    const count = this.specPortsCount;
    for (let i = 0; i < count; i++) {
      const offset = specPortRawOffset(i);
      if (this.#data[offset] === b0 && this.#data[offset + 1] === b1) {
        return i;
      }
    }
  }

  #getSpecPortRaw(index: number) {
    const offset = specPortRawOffset(index);
    return this.#data.slice(offset, offset + SPEC_PORT_ITEM_LENGTH);
  }

  findSpecPort(x: number, y: number) {
    validateCoords?.(x, y);

    const index = this.#findSpecPortOffset(x, y);
    if (index !== undefined) {
      return getSpecPortProps(this.#getSpecPortRaw(index));
    }
  }

  setSpecPort(x: number, y: number, props?: ISupaplexSpecPortProps) {
    validateCoords?.(x, y);

    let index = this.#findSpecPortOffset(x, y);
    let isNew = false;
    if (index === undefined) {
      index = this.specPortsCount;
      if (index >= SPEC_PORT_MAX_COUNT) {
        throw new RangeError("Cannot add more spec ports");
      }
      isNew = true;
    } else {
      if (!props) {
        return;
      }
    }

    const {
      setsGravity = false,
      setsFreezeZonks = false,
      setsFreezeEnemies = false,
    } = props || {};

    const [b0, b1] = specPortCoordsToOffset(x, y);
    const raw = Uint8Array.of(
      b0,
      b1,
      setsGravity ? 1 : 0,
      setsFreezeZonks ? 2 : 0,
      setsFreezeEnemies ? 1 : 0,
      0,
    );
    if (
      process.env.NODE_ENV === "production" &&
      raw.length !== SPEC_PORT_ITEM_LENGTH
    ) {
      throw new Error(
        `Invalid raw length ${raw.length}, expected ${SPEC_PORT_ITEM_LENGTH}`,
      );
    }
    this.#data.set(raw, specPortRawOffset(index));
    if (isNew) {
      this.#setSpecPortsCount(this.specPortsCount + 1);
    }
  }

  deleteSpecPort(x: number, y: number) {
    validateCoords?.(x, y);

    const index = this.#findSpecPortOffset(x, y);
    if (index !== undefined) {
      const count = this.specPortsCount;
      if (index < count - 1) {
        this.#data.set(
          this.#getSpecPortRaw(count - 1),
          specPortRawOffset(index),
        );
      }
      this.#data.set(specPortRawZeros(), specPortRawOffset(count - 1));
      this.#setSpecPortsCount(count - 1);
    }
  }
}
