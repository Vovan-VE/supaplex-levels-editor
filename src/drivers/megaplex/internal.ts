import { ILevelFooter as ISpLevelFooter } from "../supaplex/internal";

export interface IWithDemo {
  demo: Uint8Array | null;
}

export interface ILevelFooter extends ISpLevelFooter, IWithDemo {}
