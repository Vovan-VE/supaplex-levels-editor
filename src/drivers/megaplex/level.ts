import { ILevelBody } from "../supaplex/internal";
import { LevelBody } from "../supaplex/body";
import { FOOTER_BYTE_LENGTH, TITLE_LENGTH } from "../supaplex/footer";
import { isSpecPort, TILE_SPACE } from "../supaplex/tiles";
import { ISupaplexSpecPortProps } from "../supaplex/types";
import { AnyBox } from "./box";
import { LevelFooter } from "./footer";
import { IMegaplexLevel } from "./types";

const sliceFooter = (width: number, height: number, data?: Uint8Array) => {
  if (data) {
    const bodyLength = width * height;
    const minLength = bodyLength + FOOTER_BYTE_LENGTH;
    if (process.env.NODE_ENV !== "production" && data.length < minLength) {
      throw new Error(
        `Invalid buffer length ${data.length}, expected at least ${minLength}`,
      );
    }
    return data.slice(bodyLength);
  }
};

export class MegaplexLevel extends LevelFooter implements IMegaplexLevel {
  readonly #box: AnyBox;
  #body: ILevelBody;

  constructor(width: number, height: number, data?: Uint8Array) {
    super(width, sliceFooter(width, height, data));
    this.#box = new AnyBox(width, height);
    this.#body = new LevelBody(this.#box, data?.slice(0, this.#box.length));
  }

  get length() {
    return this.#box.length + super.length;
  }

  get raw() {
    const result = new Uint8Array(this.length);
    result.set(super.getRaw(this.#box.width), this.#box.length);
    result.set(this.#body.raw, 0);
    return result;
  }

  get body() {
    return this.#body;
  }

  get width() {
    return this.#box.width;
  }

  get height() {
    return this.#box.height;
  }

  get resizable() {
    return {
      maxWidth: 0x7fff,
      maxHeight: 0x7fff,
    };
  }

  copy(): this {
    return new MegaplexLevel(
      this.#box.width,
      this.#box.height,
      this.raw,
    ) as this;
  }

  resize(width: number, height: number) {
    const origSpecPorts = [...this.getSpecPorts()];

    let src = this;
    // reducing any dimension with spec port deletion
    if (
      (width < this.#box.width && origSpecPorts.some((p) => p.x >= width)) ||
      (height < this.#box.height && origSpecPorts.some((p) => p.y >= height))
    ) {
      src = this.copy();
      // delete spec ports entry
      for (let p of origSpecPorts) {
        if (
          (width < this.#box.width && p.x >= width) ||
          (height < this.#box.height && p.y >= height)
        ) {
          src = src.setTile(p.x, p.y, TILE_SPACE);
        }
      }
    }

    const temp = new Uint8Array(width * height + src.length - src.#box.length);
    temp.set(src.getRaw(width), width * height);
    let result = new MegaplexLevel(width, height, temp) as this;

    for (let y = Math.min(height, this.#box.height); y-- > 0; ) {
      for (let x = Math.min(width, this.#box.width); x-- > 0; ) {
        result = result.setTile(x, y, src.getTile(x, y));
      }
    }

    return result;
  }

  getTile(x: number, y: number) {
    return this.#body.getTile(x, y);
  }

  setTile(x: number, y: number, value: number) {
    const prev = this.#body.getTile(x, y);
    let next = this.copy();
    next.#body = next.#body.setTile(x, y, value);
    if (isSpecPort(prev)) {
      if (!isSpecPort(value)) {
        next = next.deleteSpecPort(x, y);
      }
    } else {
      if (isSpecPort(value)) {
        next = next.setSpecPort(x, y);
      }
    }
    return next;
  }

  get maxTitleLength() {
    return TITLE_LENGTH;
  }

  findSpecPort(x: number, y: number) {
    this.#box.validateCoords?.(x, y);
    return super.findSpecPort(x, y);
  }

  setSpecPort(x: number, y: number, props?: ISupaplexSpecPortProps) {
    this.#box.validateCoords?.(x, y);
    return super.setSpecPort(x, y, props);
  }

  deleteSpecPort(x: number, y: number) {
    this.#box.validateCoords?.(x, y);
    return super.deleteSpecPort(x, y);
  }
}
