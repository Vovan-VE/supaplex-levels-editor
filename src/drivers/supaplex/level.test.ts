import { createLevel, LEVEL_BYTES_LENGTH, SupaplexLevel } from "./level";
import { dumpLevel } from "./helpers.dev";
import { TILE_SP_PORT_R, TILE_SP_PORT_U, TILE_ZONK } from "./tiles-id";
import { ISupaplexSpecPort, ISupaplexSpecPortProps } from "./internal";

describe("level", () => {
  const testLevelData = Uint8Array.of(
    ...[
      "\x01",
      "\x00\x02",
      "\x00\x00\x03",
      "\x00\x00\x00\x04",
      "\x00\x00\x00\x00\x05",
      "\x00\x00\x00\x00\x00\x06",
      "\x00\x00\x00\x00\x00\x00\x07",
      "\x19\x00\x00\x00\x00\x00\x00\x08",
      "\x00\x1A\x00\x00\x00\x00\x00\x00\x09",
      "\x00\x00\x1B\x00\x00\x00\x00\x00\x00\x0A",
      "\x00\x00\x00\x1C\x00\x00\x00\x00\x00\x00\x0B",
      "\x00\x00\x00\x00\x1D\x00\x00\x00\x00\x00\x00\x0C",
      "\x00\x00\x00\x00\x00\x1E\x00\x00\x00\x00\x00\x00\x0D",
      "\x00\x00\x00\x00\x00\x00\x1F\x00\x00\x00\x00\x00\x00\x0E",
      "\x00\x00\x00\x00\x00\x00\x00\x20\x00\x00\x00\x00\x00\x00\x0F",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x21\x00\x00\x00\x00\x00\x00\x10",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x22\x00\x00\x00\x00\x00\x00\x11",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x23\x00\x00\x00\x00\x00\x00\x12",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x24\x00\x00\x00\x00\x00\x00\x13",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x25\x00\x00\x00\x00\x00\x00\x14",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x26\x00\x00\x00\x00\x00\x00\x15",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x27\x00\x00\x00\x00\x00\x00\x16",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x28\x00\x00\x00\x00\x00\x00\x17",
      "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x29\x2A\x00\x00\x00\x00\x00\x18",
    ]
      .map((s) => s.padEnd(60, "\x00").split(""))
      .flat()
      .map((c) => c.charCodeAt(0)),
    ...[0, 0, 0, 0],
    // G
    1,
    ...[0],
    // title
    ..."-- Lorem ipsum --"
      .padEnd(23)
      .split("")
      .map((c) => c.charCodeAt(0)),
    // freeze zonks
    2,
    // infotrons need
    42,
    // spec ports count
    1,
    // spec port db
    ..."\x05\xB8\x01\x02\x01\x00"
      .padEnd(10 * 6, "\x00")
      .split("")
      .map((c) => c.charCodeAt(0)),
    ...[0, 0, 0, 0],
  );
  expect(testLevelData.length).toBe(1536);

  describe("constructor", () => {
    it("no params", () => {
      expect(dumpLevel(createLevel())).toMatchSnapshot();
    });

    it("wrong size", () => {
      expect(() => createLevel(Uint8Array.of(0, 1, 6, 5))).toThrow(
        new Error(`Invalid buffer length 4, expected ${LEVEL_BYTES_LENGTH}`),
      );
    });
    it("buffer", () => {
      const level = createLevel(testLevelData);
      expect(dumpLevel(level)).toMatchSnapshot();
    });
  });

  it("copy", () => {
    const a = createLevel()
      .setTitle("First level title")
      .setTile(10, 15, 6) as SupaplexLevel;

    let b = a.copy();
    expect(dumpLevel(b)).toEqual(dumpLevel(a));

    b = b.setTitle("Copy level title").setTile(10, 15, 1);
    let dump = dumpLevel(b);
    expect(dump).not.toEqual(dumpLevel(a));
    expect(dump).toMatchSnapshot();
  });

  it("getTile", () => {
    const level = createLevel(testLevelData);

    expect(level.getTile(0, 0)).toBe(1);
    expect(level.getTile(1, 1)).toBe(2);
    expect(level.getTile(2, 2)).toBe(3);
    expect(level.getTile(23, 23)).toBe(24);

    expect(level.getTile(16, 23)).toBe(0x29);
    expect(level.getTile(17, 23)).toBe(0x2a);
  });

  describe("setTile", () => {
    it("no-op", () => {
      const level = createLevel(testLevelData);
      expect(level.setTile(0, 0, 1)).toBe(level);
      expect(level.setTile(1, 1, 2)).toBe(level);
      expect(level.setTile(23, 23, 24)).toBe(level);
    });

    it("usual", () => {
      const level = createLevel(testLevelData);
      const copy = level.setTile(10, 2, 7);
      expect(copy.getTile(10, 2)).toBe(7);
      expect(copy.specPortsCount).toBe(1);
      expect(level.getTile(10, 2)).toBe(0);
      expect(level.specPortsCount).toBe(1);
    });

    it("add spec port", () => {
      const level = createLevel(testLevelData);
      const copy = level.setTile(10, 2, TILE_SP_PORT_U);
      expect(level.getTile(10, 2)).toBe(0);
      expect(level.specPortsCount).toBe(1);
      expect(copy.getTile(10, 2)).toBe(TILE_SP_PORT_U);
      expect(copy.specPortsCount).toBe(2);
      expect([...copy.getSpecPorts()]).toEqual<ISupaplexSpecPort[]>([
        {
          x: 12,
          y: 12,
          setsGravity: true,
          setsFreezeZonks: true,
          setsFreezeEnemies: true,
        },
        {
          x: 10,
          y: 2,
          setsGravity: false,
          setsFreezeZonks: false,
          setsFreezeEnemies: false,
        },
      ]);
    });

    it("keep spec port", () => {
      const level = createLevel(testLevelData);
      const copy = level.setTile(12, 12, TILE_SP_PORT_U);
      expect(level.getTile(12, 12)).toBe(TILE_SP_PORT_R);
      expect(level.specPortsCount).toBe(1);
      expect(copy.getTile(12, 12)).toBe(TILE_SP_PORT_U);
      expect(copy.specPortsCount).toBe(1);
      expect([...copy.getSpecPorts()]).toEqual<ISupaplexSpecPort[]>([
        {
          x: 12,
          y: 12,
          setsGravity: true,
          setsFreezeZonks: true,
          setsFreezeEnemies: true,
        },
      ]);
    });

    it("remove spec port", () => {
      const level = createLevel(testLevelData);
      const copy = level.setTile(12, 12, TILE_ZONK);
      expect(level.getTile(12, 12)).toBe(TILE_SP_PORT_R);
      expect(level.specPortsCount).toBe(1);
      expect(copy.getTile(12, 12)).toBe(TILE_ZONK);
      expect(copy.specPortsCount).toBe(0);
      expect([...copy.getSpecPorts()]).toEqual([]);
    });
  });

  it("findSpecPort", () => {
    const level = createLevel(testLevelData);
    expect(level.findSpecPort(12, 12)).toEqual<ISupaplexSpecPortProps>({
      setsGravity: true,
      setsFreezeZonks: true,
      setsFreezeEnemies: true,
    });
  });

  it("setSpecPort", () => {
    const level = createLevel(testLevelData);
    expect(level.setSpecPort(12, 12)).toBe(level);
    expect(
      level.setSpecPort(12, 12, {
        setsGravity: true,
        setsFreezeZonks: true,
        setsFreezeEnemies: true,
      }),
    ).toBe(level);
    expect(
      level.setSpecPort(12, 12, {
        setsGravity: false,
        setsFreezeZonks: true,
        setsFreezeEnemies: true,
      }),
    ).not.toBe(level);
  });
});
