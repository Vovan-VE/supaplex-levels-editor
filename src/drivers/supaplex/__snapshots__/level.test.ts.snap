// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`level constructor buffer 1`] = `
Object {
  "cells": Array [
    "o                                                           ",
    " ·                                                          ",
    "  @                                                         ",
    "   &                                                        ",
    "    ▢                                                       ",
    "     ▓                                                      ",
    "      E                                                     ",
    "⚡      D                                                    ",
    " ⊏      >                                                   ",
    "  ⊐      v                                                  ",
    "   0      <                                                 ",
    "    1      ^                                                ",
    "     2      ⊃                                               ",
    "      3      ∪                                              ",
    "       ▧      ⊂                                             ",
    "        5      ∩                                            ",
    "         6      ✂                                           ",
    "          7      Y                                          ",
    "           8      T                                         ",
    "            9      R                                        ",
    "             ⊓      |                                       ",
    "              ⊔      -                                      ",
    "               ⬚      +                                     ",
    "                WW     ✧                                    ",
  ],
  "height": 24,
  "infotronsNeed": 42,
  "initialFreezeZonks": true,
  "initialGravity": true,
  "maxTitleLength": 23,
  "specialPorts": Array [
    Object {
      "setsFreezeEnemies": true,
      "setsFreezeZonks": true,
      "setsGravity": true,
      "x": 12,
      "y": 12,
    },
  ],
  "title": "-- Lorem ipsum --      ",
  "width": 60,
}
`;

exports[`level constructor no params 1`] = `
Object {
  "cells": Array [
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
    "                                                            ",
  ],
  "height": 24,
  "infotronsNeed": 0,
  "initialFreezeZonks": false,
  "initialGravity": false,
  "maxTitleLength": 23,
  "specialPorts": Array [],
  "title": "                       ",
  "width": 60,
}
`;
